applications/crossbar/test/cb_harrys_SUITE.erl

```diff
@@ -0,0 +1,187 @@
+  <<"kippy@guillain.us">> = get_value(<<"email">>, NewData1),
+  Id1 = get_value(<<"id">>, NewData1), % match id
+
+  ct:comment("There must be 2 Harrys right now"),
+  {200, Resp2} = api_call(get),
+  2 = get_value(<<"page_size">>, Resp2),
+  2 = length(get_value(<<"data">>, Resp2)),
+
+  ct:comment("Let's delete Harry1"),
+  {200, Resp3} = api_call(delete, [], Id1),
+  Data3 = get_value(<<"data">>, Resp3),
+  true = get_value(<<"deleted">>, get_value(<<"_read_only">>, Data3)),
+  Id1 = get_value(<<"id">>, Data3),
+
+  ct:comment("Trying to delete an unexisting Harry should return 404"),
+  % Harry1 was already deleted on the previous test
+  {404, Resp4} = api_call(delete, [], Id1),
+  <<"error">> = get_value(<<"status">>, Resp4),
+  <<"bad_identifier">> = get_value(<<"message">>, Resp4),
+
+  ct:comment("There must be only 1 Harry left"),
+  {200, Resp5} = api_call(get),
+  1 = get_value(<<"page_size">>, Resp5),
+  1 = length(get_value(<<"data">>, Resp5)),
+
+  {comment, ""}.
+
+%% =============================================================================
+%% Private (helpers)
+%% =============================================================================
+-spec login_user() ->
+  ok.
+login_user() ->
+  URL = "http://192.168.99.100:8000/v2/user_auth",
+  Creds = <<"1141aeae55223480eca2d9e50e3ecb3b">>,
+  ID = <<"harenson">>,
+  % JSON={"data": {"credentials": "value here", "account_name": "value here"}}.
+  ReqBody = jiffy:encode({[{<<"data">>, {[{<<"credentials">>, Creds},
+                                          {<<"account_name">>, ID}]} }]}),
+  Request = {URL, [], "application/json", ReqBody},
+  {ok, {{_Vsn, 201, _HTTPText}, _RespHeaders, RespBody}} =
+    httpc:request(put, Request, [], []),
+  DecodedResp = jiffy:decode(RespBody),
+  Data = get_value(<<"data">>, DecodedResp),
+  AccID = get_value(<<"account_id">>, Data),
+  Token = get_value(<<"auth_token">>, DecodedResp),
+  ok = application:set_env(cb_harrys_tests, account_id, binary_to_list(AccID)),
+  ok = application:set_env(cb_harrys_tests, auth_token, binary_to_list(Token)).
+
+-spec logout_user() ->
+  ok.
+logout_user() ->
+  % TODO: perform api call to logout the user.
+  ok = application:unset_env(cb_harrys_tests, account_id),
+  ok = application:unset_env(cb_harrys_tests, auth_token).
+
+-spec clean_harrys() ->
+  ok.
+clean_harrys() ->
+  {200, Resp} = api_call(get),
+  ExistingHarrys = get_value(<<"data">>, Resp),
+  ok = lists:foreach(
+         fun(Harry) ->
+             {200, _} = api_call(delete, <<>>, get_value(<<"id">>, Harry))
+         end,
+         ExistingHarrys
+        ).
+
+-spec account_id() ->
+  string().
+account_id() ->
+  {ok, AccID} = application:get_env(cb_harrys_tests, account_id),
+  AccID.
+
+-spec auth_token() ->
+  string().
+auth_token() ->
+  {ok, AuthToken} = application:get_env(cb_harrys_tests, auth_token),
+  AuthToken.
+
+-spec build_harry(FName::binary(), LName::binary(), Email::binary()) ->
+  jiffy:json_value().
```

Comment:
kz_json has a lot of exported types to use. Also, look in kz_types.hrl for lots of types we use. For instance, if the arg cannot be an empty binary (it must be non-empty), we have defined `ne_binary()` to reflect that. A weird convention if a term can be `'undefined' | type()` is to call it `api_type()`. For instance, `'undefined' | ne_binary()` is typed as `api_ne_binary()`.